Merge external infrastructure, peered VPCs, internal ELB and private subnets E2E test.

josh-ferrell opened the pull request:
Merge external infrastructure, peered VPCs, internal ELB and private subnets E2E test.

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Merges the tests for validating use of the following:
- Externally managed VPC and subnets
- Externally managed security groups
- Clusters in peered VPCs
- Clusters built with an internal ELB
- Clusters built using only private subnets

This should eliminate the flaky tests caused of reaching the VPC limit.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3360, #3371

**Special notes for your reviewer**:

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [X] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [X] adds or updates e2e tests

josh-ferrell commented:
/test pull-cluster-api-provider-aws-e2e

josh-ferrell commented:
Modified to address #3371 as well.

josh-ferrell commented:
/test pull-cluster-api-provider-aws-e2e

sedefsavas commented on test/e2e/suites/unmanaged/unmanaged_functional_test.go:
```diff
@@ -684,157 +685,25 @@ var _ = ginkgo.Context("[unmanaged] [functional]", func() {
-			ginkgo.By("Deleting route tables")
-			Expect(shared.DeleteRouteTable(e2eCtx, priRouteID)).To(BeTrue())
-			Expect(shared.DeleteRouteTable(e2eCtx, pubRouteID)).To(BeTrue())
+	ginkgo.Describe("External infrastructure, VPC peering, internal ELB and private subnet use only", func() {
```
Would be worth to mention BYO SG is also included in this test either in the name or comment.

sedefsavas commented:
Which part solves https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3371?

josh-ferrell commented:
> Which part solves #3371?

Several different parts. Since infrastructure cleanup was previously in the container node it wouldn't always get done if the management cluster creation, cluster move or workload cluster creation failed for some reason. I moved it to the setup nodes to make sure it's done every time. 

I also added functions to look for and delete orphaned instances, load balancers, security groups or security group rules. Infrastructure cleanup would previously fail if the cluster tear down wasn't successful.

sedefsavas commented on test/e2e/suites/unmanaged/unmanaged_functional_test.go:
```diff
@@ -862,6 +722,37 @@ var _ = ginkgo.Context("[unmanaged] [functional]", func() {
+
+		ginkgo.JustAfterEach(func() {
+			shared.ReleaseResources(requiredResources, config.GinkgoConfig.ParallelNode, flock.New(shared.ResourceQuotaFilePath))
+			shared.DumpSpecResourcesAndCleanup(ctx, "", namespace, e2eCtx)
```
If there is a problem with calling this upon failure here, this might be a problem in other tests as well, right?

josh-ferrell commented:
/test pull-cluster-api-provider-aws-e2e

sedefsavas commented:
> Since infrastructure cleanup was previously in the container node it wouldn't always get done if the management cluster creation, cluster move or workload cluster creation failed for some reason. I moved it to the setup nodes to make sure it's done every time.

This is interesting, so Deferred clean up does not work.
Does moving DumpSpecResourcesAndCleanup to a higher level AfterEach() and get rid of 			
`defer shared.DumpSpecResourcesAndCleanup(ctx, "", namespace, e2eCtx)` from the other tests too work?

sedefsavas commented on test/e2e/suites/unmanaged/unmanaged_functional_test.go:
```diff
@@ -862,6 +722,37 @@ var _ = ginkgo.Context("[unmanaged] [functional]", func() {
+			cPeering, _ = shared.CreatePeering(e2eCtx, mgmtClusterName+"-"+wlClusterName, *mgmtClusterInfra.VPC.VpcId, *wlClusterInfra.VPC.VpcId)
+		})
+
+		ginkgo.JustAfterEach(func() {
```
Let's add some comments to both to the test and the helpers for clarifying things like why we need JustAfter and JustBefore closures for this test specifically and what topology is created in CreateInfrastructure() method etc.

josh-ferrell replied:
Done

josh-ferrell commented:
> > Since infrastructure cleanup was previously in the container node it wouldn't always get done if the management cluster creation, cluster move or workload cluster creation failed for some reason. I moved it to the setup nodes to make sure it's done every time.
> 
> This is interesting, so Deferred clean up does not work. Does moving DumpSpecResourcesAndCleanup to a higher level AfterEach() and get rid of `defer shared.DumpSpecResourcesAndCleanup(ctx, "", namespace, e2eCtx)` from the other tests too work?

I was just starting to look into this based on your comment above. This may be related primarily to the complexity of the test. Most failures seemed to stem from a timeout but I couldn't pinpoint exactly why since it was sporadic. My primary goal here is addressing @pydctw issue but it could point to a broader problem. As much as I've run these tests locally I've seen artifacts left behind during failure like kind clusters.

sedefsavas commented:
We can track this in a separate issue and improve after merging this.

sedefsavas commented:
Also, since clean up here has problems, we should work on adding cleanup to Boskos.
Created an issue for VPC peering cleanup:
https://github.com/kubernetes-sigs/boskos/issues/120

Feel free to add issues for resources that we create in the external test that is already not here https://github.com/kubernetes-sigs/boskos/tree/master/aws-janitor/resources

sedefsavas commented:
/test pull-cluster-api-provider-aws-e2e

sedefsavas commented:
/lgtm
/approve
